pxr/base/tf/iterator.h

```diff
@@ -409,6 +409,148 @@ constexpr size_t TfArraySize(const T (&array)[N]) noexcept
     return N;
 }
 
+/// A reverse iterator adapter for `std::reverse_iterator` that provides
+/// an `operator->` compatible with proxy reference types.
+/// This should only be used when the underlying iterator's reference
+/// is a value type and should become unnecessary in newer compilers and C++20.
+template <typename UnderlyingIterator>
+class Tf_ProxyReferenceReverseIterator :
+    std::reverse_iterator<UnderlyingIterator> {
+    // private API for interacting with an STL reverse_iterator of the
+    // UnderlyingIterator
+    using ReverseIterator = std::reverse_iterator<UnderlyingIterator>;
+    const ReverseIterator& _reverse_iterator() const {  return *this; }
+    ReverseIterator& _reverse_iterator() { return *this; }
+    explicit Tf_ProxyReferenceReverseIterator(const ReverseIterator& it)
+        : ReverseIterator(it) {}
+    explicit Tf_ProxyReferenceReverseIterator(ReverseIterator&& it)
+        : ReverseIterator(it) {}
+public:
+    using iterator_type = typename ReverseIterator::iterator_type;
+    using iterator_category = typename ReverseIterator::iterator_category;
+    using value_type = typename ReverseIterator::value_type;
+    using reference = typename ReverseIterator::reference;
+    using pointer = typename ReverseIterator::pointer;
+    using difference_type = typename ReverseIterator::difference_type;
+
+    static_assert(!std::is_reference<reference>::value,
+                 "Tf_ProxyReferenceReverseIterator should only be used "
+                 "when the underlying iterator's reference type is a "
+                 "proxy (MyTypeRef) and not a true reference (MyType&)."
+                 "Use std::reverse_iterator instead.");
+    static_assert(std::is_same<iterator_category,
+                               std::random_access_iterator_tag>::value,
+                 "Tf_ProxyReferenceReverseIterator must wrap a random "
+                 "access iterator.");
+
+    Tf_ProxyReferenceReverseIterator() = default;
+    explicit Tf_ProxyReferenceReverseIterator(UnderlyingIterator it) :
+        ReverseIterator(it) {
+    }
+
+    // Operators and functions which can just use the underlying STL
+    // implementation
```

Comment:
Since these three functions are the only ones that work as is from std::reverse_iterator, I'm wondering if it would be better to use a private inherit of std::reverse_iterator and just explicitly implement the pass throughs to the parent for these functions also. The extra protection against converting to or returning as the underlying STL implementation seems beneficial.